modules/thermalsensor_generic.c

```diff
@@ -0,0 +1,1451 @@
+                thermal_sensor_generic_set_temp_func(sensor,
+                                                     tsg_util_read_temp_dC);
+            }
+            else if( !strcmp(type, "mC") ) {
+                thermal_sensor_generic_set_temp_func(sensor,
+                                                     tsg_util_read_temp_mC);
+            }
+            else {
+                dsme_log(LOG_ERR, PFIX"%s:%d: unknown/missing temp type: %s",
+                         config, line, type);
+                thermal_sensor_generic_set_temp_func(sensor, 0);
+            }
+        }
+        else if( !strcmp(key, "Meta") ) {
+            // Meta: <sensor_name> [temperature_offset]
+            char *name = tsg_util_slice_str(&pos);
+            int   offs = tsg_util_slice_int(&pos);
+            thermal_sensor_generic_set_depends_on(sensor, name);
+            thermal_sensor_generic_set_temp_offs(sensor, offs);
+        }
+        else if( !strcmp(key, "Mode") ) {
+            // Mode: <mode_write_path> <enable_string> <disable_string>
+            char *path    = tsg_util_slice_str(&pos);
+            char *enable  = tsg_util_slice_str(&pos);
+            char *disable = tsg_util_slice_str(&pos);
+            thermal_sensor_generic_set_mode_control(sensor,
+                                                    path, enable, disable);
+        }
+        else if( (rc = tsg_objects_parse_level(key)) != -1 ) {
+            // Low|Normal|...|Fatal|Invalid: <mintemp> <minwait> <maxwait>
+            int mintemp = tsg_util_slice_int(&pos);
+            int minwait = tsg_util_slice_int(&pos);
+            int maxwait = tsg_util_slice_int(&pos);
+            thermal_sensor_generic_set_limit(sensor, rc,
+                                             mintemp, minwait, maxwait);
+        }
+        else {
+            dsme_log(LOG_ERR, PFIX"%s:%d: unknown item: %s",
+                     config, line, key);
+            continue;
+        }
+
+        /* Check if line contains unparsed elements */
+        pos = tsg_util_slice_token(pos, 0, 0);
+        if( *pos != 0 && *pos != '#' ) {
+            dsme_log(LOG_WARNING, PFIX"%s:%d: excess elements: %s",
+                     config, line, pos);
+        }
+    }
+
+EXIT:
+    free(buff);
+
+    if( file ) fclose(file);
+}
+
+/** Cleanup thermal objects held in a linked list
+ *
+ * @param list    head of linked list
+ */
+static void
+tsg_objects_quit(GSList **list)
+{
+    for( GSList *item = *list; item; item = item->next ) {
+        thermal_object_t *object = item->data;
+
+        if( !object )
+            continue;
+
+        item->data = 0;
+        thermal_object_delete(object);
+    }
+
+    g_slist_free(*list), *list = 0;
+
+}
+
+/** Create linked list of thermal objects based on config files
+ *
+ * @param list    head of linked list
+ */
+static void
+tsg_objects_init(GSList **list)
+{
+    static const char pat[] = "/etc/dsme/thermal_sensor_*.conf";
+
+    glob_t gl = {};
```

Comment:
Kind of. Explicit initialization of auto structure, specifies less values than needed -> reminder are initialized to zero or null depending on type. Before c99 this was done for static data only, but IIRC gcc does it anyway even if c99 is not requested.